### DISCUSSION TITLE
Simultaneously Support Multiple Authentication Mechanisms

### DISCUSSION THREAD

**Discussion by jamlo** [nats-io/nats-server#6062] [Q&A] [answered]
Hi,

Is it possible for a NATS server to support **Token Authentication**, as well as **JWT Authentication** at the same time?

I've tried to configure the server to support both, but when **JWT Authentication** is introduced, **Token Authentication** is ignored and the server will reject any client authenticating with a token.

Thanks

**Comment by MauriceVanVeen** [comment, marked as answer]
Believe at the moment the server doesn't allow mixed modes of authentication within the same config, like combining token and JWT auth. But in essence there are two alternative ways to achieve this:

1. Use a leaf node. You connect to the leaf node using **Token Authentication**, and the leaf node is connected up to another server/cluster that  uses **JWT Authentication**. A leaf node allows you to have separate security domains, for example different authN/Z.
Link to the docs: https://docs.nats.io/running-a-nats-service/configuration/leafnodes

2. An alternative is using Auth Callout. A client connects to your server with either token or JWT (or user/pass, etc.), the server calls out to a service that you write which does the authentication and returns credentials back to the server to identify that client connection.
Link to the docs: https://docs.nats.io/running-a-nats-service/configuration/securing_nats/auth_callout

**Reply by jamlo** [reply]
Got it. Thank you @MauriceVanVeen .
